### ISSUE TITLE
Mercurial 3.2 breaks git-remote-hg

### ISSUE THREAD

**Issue by tomxtobin** [felipec/git-remote-hg#27]
Upgrading to Mercurial 3.2 appears to break git-remote-hg.

I get the following upon running `git fetch`:
```
Traceback (most recent call last):
  File "/home/tomxtobin/bin/git-remote-hg", line 1322, in <module>
    sys.exit(main(sys.argv))
  File "/home/tomxtobin/bin/git-remote-hg", line 1286, in main
    repo = get_repo(url, alias)
  File "/home/tomxtobin/bin/git-remote-hg", line 433, in get_repo
    repo.pull(peer, heads=None, force=True)
  File "/usr/lib/python2.7/site-packages/mercurial/repoview.py", line 307, in __getattr__
    return getattr(self._unfilteredrepo, attr)
AttributeError: 'localrepository' object has no attribute 'pull'
```
Downgrading to Mercurial 3.1.2 resolves the issue.

**Comment by djmitche** [comment]
With Mercuail 3.2.3 and the patch from #28 applied, I'm seeing

```
Traceback (most recent call last):
  File "/usr/bin/git-remote-hg", line 1307, in <module>
    sys.exit(main(sys.argv))
  File "/usr/bin/git-remote-hg", line 1271, in main
    repo = get_repo(url, alias)
  File "/usr/bin/git-remote-hg", line 429, in get_repo
    repo.pull(peer, heads=None, force=True)
  File "/usr/lib64/python2.7/site-packages/mercurial/repoview.py", line 307, in __getattr__
    return getattr(self._unfilteredrepo, attr)
AttributeError: 'mqrepo' object has no attribute 'pull'
```

Which is slightly different, but perhaps related?

**Comment by fingolfin** [comment]
This looks as if the patch is not actually applied -- perhaps you have multiple copies of git-remote-hg on your system and patched the wrong one?

**Comment by djmitche** [comment]
Yeah, you're right.  Ebuild error on my part.  Sorry for the noise.

**Comment by AdamWill** [comment]
Works for me on Fedora Rawhide, FWIW (I'm backporting it to the Fedora package now).

**Comment by TheKing42** [comment]
I just cloned right from github and got this:

git clone "hg::https://code.google.com/p/pychess/"
Cloning into 'pychess'...
Traceback (most recent call last):
  File "/home/chris/bin/git-remote-hg", line 1322, in <module>
    sys.exit(main(sys.argv))
  File "/home/chris/bin/git-remote-hg", line 1286, in main
    repo = get_repo(url, alias)
  File "/home/chris/bin/git-remote-hg", line 433, in get_repo
    repo.pull(peer, heads=None, force=True)
  File "/usr/lib/python2.7/site-packages/mercurial/repoview.py", line 307, in __getattr__
    return getattr(self._unfilteredrepo, attr)
AttributeError: 'localrepository' object has no attribute 'pull'

**Comment by AdamWill** [comment]
@TheKing42 the fix has not been merged yet, you have to check out the branch from https://github.com/felipec/git-remote-hg/pull/28 or apply the fix to your master checkout.

**Comment by fingolfin** [comment]
@TheKing42 you could also just use my fork, which fixes this issue and a few other minor things: https://github.com/fingolfin/git-remote-hg

**Comment by nengxu** [comment]
@fingolfin I used your fork to solve this bug. Thanks.

**Comment by bukzor** [comment]
Please to merge.

**Comment by bukzor** [comment]
You can also use mercurial<3.3 as a workaround.

**Comment by DanielGGordon** [comment]
Is @fingolfin changes not merged? I just pulled today and got the same error. Will try @fingolfin's fork

**Comment by pcn** [comment]
:+1: for @fingolfin's fork.  Is there a discussion of what needs to be done for a merge?

**Comment by fingolfin** [comment]
@pcn For a merge to happen, Felipe would have to do it, i.e. resurface. But it seems that currently he is busy with other things. Anyway, as long as my fork works for you, that should do it, right? (That said, I don't really use git-remote-hg myself very mich right now, so it indeed would be nice if there was somebody to properly maintain this project).

**Comment by fingolfin** [comment]
@DanielGGordon No it is not merged. Rest assured, if it ever gets merged, this issue will get a comment clearly stating so ;-)

**Comment by djmitche** [comment]
Worth noting that
  https://github.com/glandium/git-cinnabar
may be a useful substitute given the apparent abandonment of this project.

Dustin

**Comment by nengxu** [comment]
@djmitche git-cinnabar seems not quite active. The last commit is in April. Have you used it in real projects?

**Comment by djmitche** [comment]
I haven't as I've begun using hg directly,  but I've heard good things from
the people using it. It's possible it hasn't needed any updates in the last
5 months.

**Comment by F30** [comment]
@nengxu The 'next' branch had its last commit on 28 August. I've used Cinnabar, it works fine but doesn't support merges at the moment, which is quite a limitation.

**Comment by nengxu** [comment]
@djmitche @F30 Thanks for sharing info. I may stay with git-remote-hg for now since it works fine except this issue:

https://github.com/felipec/git-remote-hg/issues/10

**Comment by nrolland** [comment]
same error as of today

**Comment by fingolfin** [comment]
@nrolland So, have you tried my fork, as suggested?

**Comment by nrolland** [comment]
no, I just went to the one mentioned by @djmitche.. is this better to use this fork ? I find it absolutely idiotic that people would do "innovation" on the source control front, anything that makes hg get out of my way is good

**Comment by fingolfin** [comment]
Well, `git-cinnabar` is a completely different project, with its own sets of pros and cons... Anyway, if it works reasonably for you, I see no need to switch.

**Comment by nrolland** [comment]
for noobs like me you recommend this over cinnabar ?

On Fri, Oct 30, 2015 at 6:28 PM, Max Horn <notifications@github.com> wrote:

> Well, git-cinnabar is a completely different project, with its own sets
> of pros and cons... Anyway, if it works reasonably for you, I see no need
> to switch.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/felipec/git-remote-hg/issues/27#issuecomment-152594352>
> .
>



-- 
Nicolas Rolland
mobile : +33 6 62 88 42 92

**Comment by xrchz** [comment]
@felipec how come @fingolfin's patch is not merged yet?

**Comment by alexei-31** [comment]
@fingolfin is not patch, its actual fork against this, which depricated.
Author of @felipec is not update something here since 3 Jun 2014.

**Comment by fingolfin** [comment]
@nrolland sorry, didn't see your last question. I have never tried `git-cinnabar`, and so I can neither recommend for nor against it. My understanding is that using it is about exactly as easy / difficult as using this project.

What I *can* say is that my fork of this project has been working well for me for some years now. However, that doesn't mean it will work perfectly for you -- perhaps you will use it quite differently from the way I use it, uncovering problems I never noticed myself... who knows...

**Comment by xrchz** [comment]
@fingolfin I'm using your fork now, and it's working great :) Thanks

**Comment by felipec** [comment]
Fixed in 1e27907.

**Comment by felipec** [comment]
The home page clearly states:

Contributing

Send your patches to the mailing list git-fc@googlegroups.com (no need to subscribe).

Nobody sent any patches.
